### PR TITLE
Add DQ ranking statistic to coinc search workflow

### PR DESCRIPTION
**Pull request by dethodav**
This pull request is still work in progress as it depends on #3804 and includes commits that will be added as part of that PR.

This PR adds support for the new DQ ranking statistic to be used in the coinc search workflow. This required a number of changes to `pycbc_make_coinc_search_workflow` to support the new executables and workflow structure. However, this new code should be fully back-compatible.

### CODE REVIEW

**Comment by titodalcanton** [comment]
@dethodav as we are close to merging #3804, please make sure this is appropriately rebased. At the moment it conflicts with master and appears to introduce many of the things already introduced in #3804.

**Comment by titodalcanton** [comment]
What is the status of this? Is it still relevant?

**Comment by titodalcanton** [comment]
@maxtrevor you may actually be able to push to Derek's branch. Give it a try, if it does not work you can create a new branch, though I would prefer to continue to use this PR for cleanliness.

**Comment by dethodav** [comment]
@titodalcanton @maxtrevor I can take a look at this again - apologies for letting this sit for so long

**Comment by dethodav** [comment]
@titodalcanton @maxtrevor I've gone ahead and rebased to the current master branch and implement some of the previous suggestions by @titodalcanton. I'm currently rerunning a full workflow to confirm that everything is still working after the rebase.

**Comment by maxtrevor** [comment]
@titodalcanton @dethodav whats next to keep this moving?

File: pycbc/workflow/dq.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            output += new_node.output_files
+    elif dq_type == 'flag':
+        flag_str = workflow.cp.get_opt_tags("workflow-data_quality",
+                                            'flag-name', dq_tags)
+        ifo = flag_str[:2]
+        ifo_insp = [insp for insp in insps if (insp.ifo == ifo)]
+        assert len(ifo_insp)==1
+        ifo_insp = ifo_insp[0]
+        flag_name = flag_str
+        logging.info("Creating job for flag %s" % (flag_name))
```
Comment (titodalcanton) [inline pycbc/workflow/dq.py:167]:
```suggestion
        logging.info("Creating job for flag %s", flag_name)
```

File: bin/all_sky_search/pycbc_combine_statmap [hunk 2]
```diff
@@ -57,9 +57,12 @@ for fi in files:
     ends = fi['segments/{}/end'.format(key)][:]
     indiv_segs[key] = pycbc.events.veto.start_end_to_segments(starts, ends)
 
-# Convert segmentlistdict to a list ('seglists') of segmentlists
-# then numpy.sum(seglists, axis=0) does seglists[0] + seglists[1] + ...
-foreground_segs = numpy.sum(list(indiv_segs.values()), axis=0)
+if len(indiv_segs.values())==1:
```
Comment (titodalcanton) [inline bin/all_sky_search/pycbc_combine_statmap:60]:
```suggestion
if len(indiv_segs.values()) == 1:
```

File: bin/plotting/pycbc_plot_bank_bins [hunk 3]
```diff
@@ -91,12 +97,12 @@ parser.add_argument('--log-y', action='store_true',
 args = parser.parse_args()
 
 bank = H5BankFile(args.bank_file, 'r')
-bank.f_lower = args.f_lower
+f_lower  = args.f_lower or bank['f_lower'][:]
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_bank_bins:100]:
```suggestion
f_lower = args.f_lower or bank['f_lower'][:]
```

File: bin/plotting/pycbc_plot_dq_likelihood_vs_time [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import h5py
+from matplotlib import use; use('Agg')
+from matplotlib import pyplot
+from pycbc.version import git_verbose_msg as version
+import pycbc.results
+
+parser = argparse.ArgumentParser(description=__doc__)
+parser.add_argument('--version', action='version', version=version)
+parser.add_argument('--verbose', action="store_true")
+parser.add_argument("--ifo", type=str,required=True)
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_dq_likelihood_vs_time:18]:
```suggestion
parser.add_argument("--ifo", type=str, required=True)
```

File: bin/plotting/pycbc_plot_dq_likelihood_vs_time [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+pycbc.init_logging(args.verbose)
+
+
+ifo = args.ifo
+
+f = h5py.File(args.dq_file, 'r')
+
+bin_name = args.background_bin
+
+if bin_name not in f['%s/dq_vals'%ifo].keys():
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_dq_likelihood_vs_time:33]:
```suggestion
if bin_name not in f[f'{ifo}/dq_vals'].keys():
```

File: bin/plotting/pycbc_plot_dq_likelihood_vs_time [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    raise ValueError("Background bin name not found in DQ file")
+else:
+    yvals = f['%s/dq_vals/%s'%(ifo,bin_name)][:]
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_dq_likelihood_vs_time:36]: No need for the `else` block here, the code just carries on.
```suggestion
yvals = f[f'{ifo}/dq_vals/{bin_name}'][:]
```

File: bin/plotting/pycbc_plot_dq_likelihood_vs_time [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+xvals = f['%s/times'%ifo][:]
```
Comment (titodalcanton) [inline bin/plotting/pycbc_plot_dq_likelihood_vs_time:38]:
```suggestion
xvals = f[f'{ifo}/times'][:]
```

File: bin/workflows/pycbc_make_coinc_search_workflow [hunk 8]
```diff
@@ -494,6 +530,26 @@ for key in final_bg_files:
     snrifar_summ += closed_page
     layout.two_column_layout(closed_dir, closed_page)
 
+#################### Plotting of data quality results #########################
+
+# DQ log likelihood plots
+for dqf,dql in zip(dqfiles,dqfile_labels):
+    outdir = rdir['data_quality/%s_trigger_rates_%s' % (dqf.ifo,dql)]
```
Comment (titodalcanton) [inline bin/workflows/pycbc_make_coinc_search_workflow:537]:
```suggestion
    outdir = rdir[f'data_quality/{dqf.ifo}_trigger_rates_{dql}']
```

File: bin/workflows/pycbc_make_coinc_search_workflow [hunk 9]
```diff
@@ -494,6 +530,26 @@ for key in final_bg_files:
...
+    wf.make_dq_trigger_rate_plot(workflow, [dqf], outdir, tags=[dql,'full_data'])
+
+    outdir = rdir['data_quality/%s_percentiles_%s' % (dqf.ifo,dql)]
```
Comment (titodalcanton) [inline bin/workflows/pycbc_make_coinc_search_workflow:540]:
```suggestion
    outdir = rdir[f'data_quality/{dqf.ifo}_percentiles_{dql}']
```

File: pycbc/workflow/dq.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        node.add_opt('--ifo', ifo)
+        node.add_input_opt('--bank-file', bank_file)
+        node.add_input_opt('--trig-file', trig_file)
+        node.add_input_list_opt('--dq-file', dq_files)
+        node.new_output_file_opt(workflow.analysis_time,'.hdf',
+                                 '--output-file')
+        return node
+
+class PyCBCCalculateDQFlagExecutable(Executable):
+    # current_retention_level = Executable.ALL_TRIGGERS
```
Comment (titodalcanton) [inline pycbc/workflow/dq.py:69]: Any reason not to remove this line?

**Comment by titodalcanton** [comment]
I just left a bunch of minor style suggestions. As a more general comment, the two plotting scripts added by this PR look very similar, does it make sense to consolidate them into a single script?

@dethodav was also running one more test to check everything is ok, I would like to hear from that before approving.

**Comment by dethodav** [comment]
@titodalcanton My rerun of the workflow after rebasing was successful: 

https://ldas-jobs.ligo.caltech.edu/~derek.davis/cbc/preO4/pycbc_idq/merge_request_tests/dq-workflow-simple-gwosc-chunkgwosc-1/

I did identify an issue where one of the plotting scripts can fail when small amounts of data are analyzed. I fixed this separately in this PR: https://github.com/gwastro/pycbc/pull/4153

I'll work on addressing these comments and other style suggestions as soon as I can.

**Comment by maxtrevor** [comment]
@dethodav Any updates on this?

**Comment by titodalcanton** [comment]
Most of my comments above can simply be accepted by clicking the "commit suggestion" button, so should be fairly quick to address.

**Comment by maxtrevor** [comment]
@dethodav any progress on this? We really need to get this merged in ASAP

**Comment by titodalcanton** [comment]
I took the liberty of applying most of my suggestions, as they were really trivial. I think a rebase on master is needed now to make the CI tests pass.

**Comment by dethodav** [comment]
@titodalcanton @maxtrevor I've finally updated they PR - let me know if there's any other changes I should make!

**Review by titodalcanton — APPROVED** [review]
Thanks @dethodav @maxtrevor. I think it is time to merge this.

**Pull request merged**